### ISSUE TITLE
🎯 Adaptive Testing Engine: AI-Powered Question Selection

### ISSUE THREAD

**Issue by amolsr** [amolsr/pariksha#18]
### Overview
Implement an AI-powered adaptive testing system that dynamically adjusts question difficulty based on student performance, providing personalized assessment experiences and more accurate evaluation of student capabilities.

### Current Limitations
- Fixed question selection from predefined categories
- No difficulty adjustment based on performance
- One-size-fits-all approach to testing
- Limited assessment accuracy

#### 1. **Dynamic Difficulty Adjustment**
- **Performance-Based Selection**: Choose questions based on previous answers
- **Difficulty Scaling**: Increase/decrease difficulty based on accuracy
- **Confidence Intervals**: Adjust based on answer confidence
- **Learning Curve**: Adapt to individual learning patterns

#### 2. **Intelligent Question Selection**
- **Item Response Theory (IRT)**: Use IRT models for question selection
- **Cognitive Load Management**: Balance question complexity
- **Knowledge Gap Analysis**: Identify and target weak areas
- **Optimal Stopping**: Determine when to end the test

#### 3. **Personalized Assessment**
- **Individual Pacing**: Allow students to progress at their own pace
- **Skill-Based Routing**: Direct students to appropriate difficulty levels
- **Competency Mapping**: Map questions to specific competencies
- **Learning Objectives**: Align questions with learning goals

#### Adaptive Algorithm
```javascript
// Adaptive testing engine
class AdaptiveTestingEngine {
  constructor() {
    this.abilityEstimate = 0.5; // Initial ability estimate (0-1)
    this.confidenceInterval = 0.1;
    this.questionBank = [];
    this.answeredQuestions = [];
  }
  
  selectNextQuestion() {
    // Calculate optimal question based on current ability estimate
    const optimalQuestion = this.calculateOptimalQuestion();
    return optimalQuestion;
  }
  
  updateAbilityEstimate(question, response) {
    // Update ability estimate using IRT
    const difficulty = question.difficulty;
    const discrimination = question.discrimination;
    const guessing = question.guessing;
    
    // Use 3-parameter logistic model
    const probability = this.calculateProbability(
      this.abilityEstimate, 
      difficulty, 
      discrimination, 
      guessing
    );
    
    // Update ability estimate using maximum likelihood estimation
    this.abilityEstimate = this.maximumLikelihoodEstimation(
      this.answeredQuestions,
      this.abilityEstimate
    );
  }
  
  calculateOptimalQuestion() {
    // Find question with maximum information at current ability level
    return this.questionBank.reduce((best, question) => {
      const information = this.calculateInformation(
        this.abilityEstimate,
        question.difficulty,
        question.discrimination
      );
      
      return information > best.information ? question : best;
    });
  }
}
```

#### Question Difficulty Modeling
```javascript
// Question difficulty and discrimination parameters
const QuestionModel = {
  difficulty: Number, // -3 to +3 (easy to hard)
  discrimination: Number, // 0 to 2 (how well it discriminates)
  guessing: Number, // 0 to 1 (probability of guessing correctly)
  category: String,
  learningObjective: String,
  cognitiveLevel: String // 'remember', 'understand', 'apply', 'analyze', 'evaluate', 'create'
};
```

#### IRT Implementation
```javascript
// Item Response Theory calculations
const IRTCalculator = {
  // 3-Parameter Logistic Model
  calculateProbability: (ability, difficulty, discrimination, guessing) => {
    const exponent = discrimination * (ability - difficulty);
    return guessing + (1 - guessing) / (1 + Math.exp(-exponent));
  },
  
  // Information function
  calculateInformation: (ability, difficulty, discrimination, guessing) => {
    const p = this.calculateProbability(ability, difficulty, discrimination, guessing);
    const q = 1 - p;
    const d = discrimination;
    const c = guessing;
    
    return (d * d * (p - c) * (p - c)) / (p * (1 - c) * (1 - c));
  },
  
  // Maximum Likelihood Estimation
  maximumLikelihoodEstimation: (responses, initialAbility) => {
    // Implement MLE algorithm
    let ability = initialAbility;
    let tolerance = 0.001;
    let maxIterations = 50;
    
    for (let i = 0; i < maxIterations; i++) {
      const gradient = this.calculateGradient(ability, responses);
      const hessian = this.calculateHessian(ability, responses);
      
      const newAbility = ability - gradient / hessian;
      
      if (Math.abs(newAbility - ability) < tolerance) {
        break;
      }
      
      ability = newAbility;
    }
    
    return ability;
  }
};
```

### Database Schema Updates
```javascript
// Enhanced question model with IRT parameters
const Question = {
  // ... existing fields
  difficulty: Number, // IRT difficulty parameter
  discrimination: Number, // IRT discrimination parameter
  guessing: Number, // IRT guessing parameter
  cognitiveLevel: String, // Bloom's taxonomy level
  learningObjective: String, // Associated learning objective
  competency: String, // Required competency
  estimatedTime: Number, // Estimated time to answer
  tags: [String], // Additional categorization tags
  usageCount: Number, // How many times used
  averageScore: Number, // Average score when used
  lastUsed: Date // When last used
};

// Adaptive test session
const AdaptiveTestSession = {
  userId: ObjectId,
  testId: ObjectId,
  abilityEstimate: Number,
  confidenceInterval: Number,
  questionsAnswered: [{
    questionId: ObjectId,
    response: String,
    correct: Boolean,
    timeSpent: Number,
    timestamp: Date
  }],
  currentQuestion: ObjectId,
  status: String, // 'active', 'completed', 'paused'
  startTime: Date,
  endTime: Date
};
```

### API Endpoints
```javascript
// Adaptive testing endpoints
router.get('/adaptive/start/:testId', adaptiveController.startAdaptiveTest);
router.get('/adaptive/next-question', adaptiveController.getNextQuestion);
router.post('/adaptive/submit-answer', adaptiveController.submitAnswer);
router.get('/adaptive/ability-estimate', adaptiveController.getAbilityEstimate);
router.post('/adaptive/end-test', adaptiveController.endAdaptiveTest);
```

### Frontend Implementation
```javascript
// Adaptive test component
const AdaptiveTest = () => {
  const [currentQuestion, setCurrentQuestion] = useState(null);
  const [abilityEstimate, setAbilityEstimate] = useState(0.5);
  const [testSession, setTestSession] = useState(null);
  
  const getNextQuestion = async () => {
    const response = await fetch('/api/adaptive/next-question');
    const question = await response.json();
    setCurrentQuestion(question);
  };
  
  const submitAnswer = async (answer) => {
    const response = await fetch('/api/adaptive/submit-answer', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ answer })
    });
    
    const result = await response.json();
    setAbilityEstimate(result.abilityEstimate);
    
    // Get next question
    await getNextQuestion();
  };
  
  return (
    <div className="adaptive-test">
      <AbilityIndicator level={abilityEstimate} />
      <QuestionDisplay question={currentQuestion} onSubmit={submitAnswer} />
      <ProgressIndicator session={testSession} />
    </div>
  );
};
```

### Machine Learning Integration
```javascript
// ML model for question recommendation
const QuestionRecommendationModel = {
  train: async (historicalData) => {
    // Train model on historical test data
    const model = await tf.sequential({
      layers: [
        tf.layers.dense({ inputShape: [10], units: 64, activation: 'relu' }),
        tf.layers.dense({ units: 32, activation: 'relu' }),
        tf.layers.dense({ units: 1, activation: 'sigmoid' })
      ]
    });
    
    model.compile({
      optimizer: 'adam',
      loss: 'binaryCrossentropy',
      metrics: ['accuracy']
    });
    
    return model;
  },
  
  predict: async (model, questionFeatures) => {
    // Predict question suitability
    const prediction = model.predict(questionFeatures);
    return prediction.dataSync()[0];
  }
};
```

### Performance Optimization
- **Caching**: Cache ability estimates and question selections
- **Precomputation**: Precompute question information values
- **Batch Processing**: Process multiple students simultaneously
- **Database Indexing**: Optimize queries for question selection

### Analytics and Reporting
```javascript
// Adaptive testing analytics
const AdaptiveAnalytics = {
  getAbilityDistribution: () => {
    // Analyze ability distribution across students
  },
  
  getQuestionEffectiveness: () => {
    // Analyze question effectiveness
  },
  
  getLearningProgress: (userId) => {
    // Track individual learning progress
  },
  
  getCompetencyMapping: () => {
    // Map competencies to questions
  }
};
```

### Priority
**MEDIUM** - Significant educational value

### Acceptance Criteria
- [ ] IRT-based question selection
- [ ] Dynamic difficulty adjustment
- [ ] Real-time ability estimation
- [ ] Personalized question routing
- [ ] Competency-based assessment
- [ ] Performance analytics
- [ ] Machine learning integration
- [ ] User-friendly interface

### Dependencies
- TensorFlow.js for machine learning
- Statistical libraries for IRT calculations
- Enhanced question database schema
- Real-time data processing

### Timeline
- **Phase 1**: Basic IRT implementation (4-6 weeks)
- **Phase 2**: Adaptive algorithm development (6-8 weeks)
- **Phase 3**: ML integration and optimization (4-6 weeks)
- **Phase 4**: Analytics and reporting (3-4 weeks)

---
**Labels**: `ai`, `adaptive-testing`, `machine-learning`, `enhancement`, `education`
